### PR TITLE
Add Favorite button to the Direct connect tab

### PR DESCRIPTION
**Pull request by purifiedfr**
Yet unfinished:
* Localization ~~is not yet implemented tested (I'd appreciate help with this, I don't want to mess anything up)~~ seems to work for English (implemented under `ui.multiplayer.favorite`)
* The button should give feedback when one clicks it
* Custom favorited servers currently show "undefined" for the map (as shown below), further testing is needed

## Screenshots / POCs:
image
image

### CODE REVIEW

**Comment by WiserTixx** [comment]
Regarding your points:
- The custom favorited servers show the map as undefined because the client doesn't know anything about the server until it connects, however this is being worked on. There's nothing that can be done about that in this PR.
- The localizations looks good, however it may be clearer to the user if it says something like "Save as favorite". Once this is merged it can be translated into more languages later.
- For the visual feedback maybe show a toast message? `toastr.info`

Codewise it looks good mostly.
The empty `ng-class=""` should be able to be removed.
And could the core functionality of `vm.addCustomServer` and `vm.directConnectFavorite` maybe be merged into a function (`addCustomFavorite` or something)? Then call it via `directConnectFavorite` and `dialogFavorite`.

File: mp_locales/en-US.json [hunk 1]
```diff
@@ -93,6 +93,9 @@
     "ui.multiplayer.official": "Official Servers",
     "ui.multiplayer.servers": "Public Servers",
     "ui.multiplayer.favorites": "Favorites",
+    "ui.multiplayer.favorite": "Save as Favorite",
+    "ui.multiplayer.favorited_toast_title": "Favorited server",
+    "ui.multiplayer.favorited_toast_description": "This server has been favorited.",
```
Comment (WiserTixx) [inline mp_locales/en-US.json]:
```suggestion
    "ui.multiplayer.favoritedToast.title": "Favorited server",
    "ui.multiplayer.favoritedToast.description": "This server has been favorited.",
```
Suggestion to keep the casing consistent

File: ui/modules/multiplayer/multiplayer.js [hunk 2]
```diff
@@ -286,6 +286,22 @@ function($scope, $state, $timeout, $mdDialog, ConfirmationDialog) {
...
+		var valid = (ip.length > 0) && (port.length > 0) && !isNaN(port)
+		if (!valid) return;
+		var name = ip + ":" + port;
+		var server = {
+			cversion: await getLauncherVersion(), ip: ip, location: "--", map: "", maxplayers: "0", players: "0",
+			owner: "", playersList: "", sdesc: "", sname: name, strippedName: name,
+			custom: true, port: port
+		}
+		addFav(server)
+		toastr.info($filter('translate')('ui.multiplayer.favorited_toast_description'), $filter('translate')('ui.multiplayer.favorited_toast_title'));
```
Comment (WiserTixx) [inline ui/modules/multiplayer/multiplayer.js]:
```suggestion
		toastr.info($filter('translate')('ui.multiplayer.favoritedToast.description'), $filter('translate')('ui.multiplayer.favoritedToast.title'));
```

File: ui/modules/multiplayer/multiplayer.js [hunk 3]
```diff
@@ -286,6 +286,22 @@ function($scope, $state, $timeout, $mdDialog, ConfirmationDialog) {
...
+
+		var valid = (ip.length > 0) && (port.length > 0) && !isNaN(port)
+		if (!valid) return;
+		var name = ip + ":" + port;
+		var server = {
+			cversion: await getLauncherVersion(), ip: ip, location: "--", map: "", maxplayers: "0", players: "0",
+			owner: "", playersList: "", sdesc: "", sname: name, strippedName: name,
+			custom: true, port: port
+		}
+		addFav(server)
```
Comment (WiserTixx) [inline ui/modules/multiplayer/multiplayer.js:302]:
```suggestion
		favorites = await getFavorites();
		addFav(server)
```
This prevents the favorites being cleared when you add a server but haven't loaded the favorites tab yet.

**Review by WiserTixx — CHANGES_REQUESTED** [review]

**Comment by purifiedfr** [comment]
Thank you for these changes

**Comment by WiserTixx** [comment]
If you're ready for this PR to be reviewed, please mark it so.

**Comment by purifiedfr** [comment]
It's ready for review

**Review by WiserTixx — APPROVED** [review]

**Pull request merged**